### DISCUSSION TITLE
Add "Extend Script" button to Scene Tree Dock

### DISCUSSION THREAD

**Discussion by afterlightgames** [godotengine/godot-proposals#11289] [Editor]
When working with instanced scenes and custom classes that have scripts attached, I often want to extend the attached script to add functionality to a specific instance or create a subclass that extends it. Personally, I want to do this more often than detaching the script, but only the detach script button is displayed at the top of the dock when these nodes are selected, requiring users to right click the node to reveal the extend script button.

I propose an option to show the extend and detach script buttons side by side at the top of the scene tree dock when an instanced scene or custom node with an attached script is selected. It's a small change but it would likely be simple to implement and could improve the user experience as I imagine many users encounter fewer scenarios where they need to detach a script from an instanced scene or custom class instead of extend it, and it could improve workflows that involve extending scripts often.

![Illustration](https://github.com/user-attachments/assets/8e329138-0df9-4642-9d1d-95d7360da6f8)

**Comment by tetrapod00** [comment]
I don't think that it's very clear that `Extend Script` is used *more* than `Detach Script` (I personally don't ever use `Extend`, but I do use `Detach` often enough). So I'm not sure that changing the default here is justified.

Assuming there is some support for `Extend` being a more common workflow than `Detach` I do think either of the following could be good options:
- Show both buttons side by side. The (slight) problem with this is that it uses up more of the scarce horizontal space in the editor. Maybe at some point the toolbar at the top of the scene tab becomes two lines tall, and this isn't actually a problem.
- Show only one button, but add an editor setting to change which action is shown. This would be a good compromise if there is a lot of demand for `Extend` being more common than `Detach`, but it isn't a universal preference. There is still a cost to adding another editor setting, and for something small like this, users might never think to *search* for the editor setting, because even if they do prefer `Extend` to `Detach`, it's only a small annoyance.

I think we should see how much demand there is for this, and how common it is to prefer one or the other in typical workflows.

**Reply by afterlightgames** [reply]
Both very good solutions. I especially second the possibility that this could be provided as an editor option in order to cater to different workflows.

**Reply by Flynsarmy** [reply]
Would prefer both being shown as extending scripts is very common if you're using addons like [Godot State Charts](https://godotengine.org/asset-library/asset/1778) or adding custom functionality to inherited scenes.

We don't need editor options for this type of thing - if an option was added for every tiny UI change such as this it'd be so overly cluttered you'd never be able to find anything.

**Reply by HanzCEO** [reply]
Custom node that does not show "Extend Script" button is something that should not happen.

To show both "Extend" and "Detach" buttons is a separate feature request, not an issue at the moment.

**Comment by Flynsarmy** [comment]
I did a bit of digging, and 'Extend Script' button is actually already in the dock, just hidden. With a bit of tweaking I made it visible:
![image](https://github.com/user-attachments/assets/0a509912-8c45-433c-97ca-4edc1820ab6f)

I'm not smart enough to understand exactly why it's not showing under normal conditions, but here's the code that determines whether or not it displays:

https://github.com/godotengine/godot/blob/1f47e4c4e3a09a422e96880a7918d986dd575a63/editor/scene_tree_dock.cpp#L2802-L2810

Looks bugged honestly but we'd need someone else to take a look and confirm.

**Reply by afterlightgames** [reply]
Good find! I can't seem to get the button to show up at all so it does seem like a bug. I'll open an issue on it.

**Reply by Flynsarmy** [reply]
To keep things nice and linked up, here is the issue mentioned above https://github.com/godotengine/godot/issues/100074

**Comment by afterlightgames** [comment]
It appears that as of dev7 the issue with custom nodes not having the extend script button has been fixed. However, I would love to see this feature also extended to instanced scenes that have scripts attached!
